File: data structures and algorithms/python/06_027.py
def adaptive_heuristic(a, b, grid_type):
    if grid_type == "manhattan":
        return abs(a[0] - b[0]) + abs(a[1] - b[1])
    elif grid_type == "chebyshev":  # Allows diagonal movement
        return max(abs(a[0] - b[0]), abs(a[1] - b[1]))

File: data structures and algorithms/cs/06_027.cs
/**
 * Select appropriate heuristic based on grid movement type
 */
public static double AdaptiveHeuristic(int[] a, int[] b, string gridType)
{
   // For grid with only cardinal directions (up/down/left/right)
   if (gridType == "manhattan")
   {
      return Math.Abs(a[0] - b[0]) + Math.Abs(a[1] - b[1]);
   }
   // For grid allowing diagonal movement (8 directions)
   else if (gridType == "chebyshev")
   {
      return Math.Max(Math.Abs(a[0] - b[0]), Math.Abs(a[1] - b[1]));
   }
   // Default case (no heuristic)
   return 0;
}
